src/dparse/parser.d

```diff
@@ -3024,26 +3024,94 @@ class Parser
             advance();
             return node;
         }
-        else if (currentIs(tok!"{"))
-            mixin(parseNodeQ!(`node.blockStatement`, `BlockStatement`));
-        else if (currentIsOneOf(tok!"in", tok!"out", tok!"do") || current.text == "body")
+
+        StackBuffer inStatements;
+        StackBuffer outStatements;
+        bool requireDo;
+
+        while (currentIsOneOf(tok!"in", tok!"out"))
         {
+            const(Token)* next = peek();
+            if (next is null)
+                return null;
+
             if (currentIs(tok!"in"))
             {
-                mixin(parseNodeQ!(`node.inStatement`, `InStatement`));
-                if (currentIs(tok!"out"))
-                    mixin(parseNodeQ!(`node.outStatement`, `OutStatement`));
+                if (next.type == tok!"{")
+                {
+                    requireDo = true;
+                    if (auto s = parseInStatement())
+                        inStatements.put(s);
+                    else return null;
+                }
+                else if (next.type == tok!"(")
+                {
+                    advance();
+                    advance();
+                    if (InStatement s = parseContractExpression!InStatement())
+                        inStatements.put(s);
+                    else return null;
+                }
+                else return null;
             }
             else if (currentIs(tok!"out"))
             {
-                mixin(parseNodeQ!(`node.outStatement`, `OutStatement`));
-                if (currentIs(tok!"in"))
-                    mixin(parseNodeQ!(`node.inStatement`, `InStatement`));
+                if (next.type == tok!"{" || (index < tokens.length - 3  &&
+                    tokens[index + 2].type == tok!("identifier") &&
+                    tokens[index + 3].type == tok!(")")))
```

Comment:
I really should add a `peekIs` overload that takes an offset to avoid verbose code like this.

Reply:
Wrote https://github.com/dlang-community/libdparse/issues/246 as a reminder for later. This isn't something that needs to be changed for this pull request to be merged.